nui/src/main/java/org/terasology/nui/widgets/types/builtin/util/FieldsWidgetBuilder.java

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+package org.terasology.nui.widgets.types.builtin.util;
+
+import com.google.common.collect.Maps;
+import org.reflections.ReflectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.nui.UIWidget;
+import org.terasology.nui.WidgetUtil;
+import org.terasology.nui.databinding.Binding;
+import org.terasology.nui.widgets.types.TypeWidgetLibrary;
+import org.terasology.reflection.ReflectionUtil;
+import org.terasology.reflection.TypeInfo;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.terasology.nui.widgets.types.TypeWidgetFactory.LABEL_WIDGET_ID;
+
+public class FieldsWidgetBuilder<T> {
```

Comment:
**thought:** Also, would be nice to automatically bake in <kbd>tab</kbd> navigation here.